File: 0054-spiral-matrix/0054-spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let res = [];
    let m = matrix.length;
    let n = matrix[0].length;
    
    let top = 0;
    let bottom = m - 1;
    let left = 0;
    let right = n - 1;
    
    while(true) {
        // Traverse from left to right.
        for (let i = left; i <= right; i++) res.push(matrix[top][i]);
        if (++top > bottom) break;

        // Traverse from top to bottom.
        for (let i = top; i <= bottom; i++) res.push(matrix[i][right]);
        if (--right < left) break;

        // Traverse from right to left.
        for (let i = right; i >= left; i--) res.push(matrix[bottom][i]);
        if (--bottom < top) break;

        // Traverse from bottom to top.
        for (let i = bottom; i >= top; i--) res.push(matrix[i][left]);
        if (++left > right) break;
    }
    
    return res;
};

File: 0054-spiral-matrix/0054-spiral-matrix.rs
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        if matrix.is_empty() || matrix[0].is_empty() {
            return vec![];
        }
        
        let (mut top, mut bottom, mut left, mut right) = (0, matrix.len() - 1, 0, matrix[0].len() - 1);
        let mut res = Vec::with_capacity((bottom + 1) * (right + 1));

        loop {
            // Traverse from left to right.
            for i in left..=right {
                res.push(matrix[top][i]);
            }
            top += 1;
            if top > bottom {
                break;
            }

            // Traverse from top to bottom.
            for i in top..=bottom {
                res.push(matrix[i][right]);
            }
            if right == 0 || left == right {
                break;
            }
            right -= 1;

            // Traverse from right to left.
            for i in (left..=right).rev() {
                res.push(matrix[bottom][i]);
            }
            if bottom == 0 || top == bottom {
                break;
            }
            bottom -= 1;

            // Traverse from bottom to top.
            for i in (top..=bottom).rev() {
                res.push(matrix[i][left]);
            }
            left += 1;
            if left > right {
                break;
            }
        }
        res
    }
}
